File: leetcode/medium/Q55_JumpGame.py
from typing import List

class Solution:
    visit = []
    def canJump(self, nums: List[int]) -> bool:
        self.visit = [False] * len(nums)
        return self.jump(nums, 0)

    def jump(self, nums: List[int], idx: int):
        if idx > len(nums) - 1:
            return False

        self.visit[idx] = True
        if self.visit[len(nums)-1]:
            return True
        if self.visit[idx+min(nums[idx], len(nums)-idx-1)]:
            return False
        b = False
        for i in range(min(nums[idx], len(nums)-idx-1), 0, -1):
            if not self.visit[idx + i]:
                b = b | self.jump(nums, idx + i)
            if self.visit[len(nums) - 1]:
                return b
        return b

File: src/medium/Q55_JumpGame.java
package medium;

public class Q55_JumpGame {
    static class Solution {
        public boolean canJump(int[] nums) {
            int length = nums.length;
            int idx = 0;

            while(idx + nums[idx] < length -1){
                int nextIdx = idx;
                int maxJump = 0;
                for (int i = 1; i <= nums[idx] && idx + i < length; i++) {
                    if(maxJump < idx + nums[idx+i] + i){
                        maxJump = idx + nums[idx+i] + i;
                        nextIdx = idx + i;
                    }
                }
                if(nextIdx == idx) return false;
                idx = nextIdx;
            }
            return true;
        }
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        int[] nums = {2, 3, 1, 1, 4};
        s.canJump(nums);
    }
}
